### ISSUE TITLE
[Bug]: Create collection hangs and cannot create producer on topic with backlog quota exceeded

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#16114]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20220316-d4ad785b
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2): 2.0.2.dev5
- OS(Ubuntu or CentOS): CentOS
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. deploy milvus cluster by operator, and all replicas are 1
2. Run the following script multiple times: create a collection and insert 10 million vectors in multiples.
However, during this period, the disk of the node was filled many times, causing many pods to be evicted and restarted.
And I'm not sure if I ever dropped the collection.
```
        host = "10.100.32.xxx"
        port = 31746
        hdf5_source_file = "/Users/nausicca/Downloads/vectors/sift-128-euclidean.hdf5"
        ni = 50000
        nb = 1000000      
        # define field and schema
        collection_w = ApiCollectionWrapper()
        filed_w = ApiFieldSchemaWrapper()
        schema_w = ApiCollectionSchemaWrapper()
        fields = [filed_w.init_field_schema(name="id", dtype=DataType.INT64, is_primary=True)[0],
                  filed_w.init_field_schema(name="vec", dtype=DataType.FLOAT_VECTOR, dim=128)[0]]
        schema = schema_w.init_collection_schema(fields, auto_id=True)[0]

        # create collection
        collection_w.init_collection(name=cf.gen_unique_str("disk"), schema=schema, shards_num=1, timeout=20)
        log.info(collection_w.num_entities)

        dataset = h5py.File(hdf5_source_file)

        # insert
        vectors = np.array(dataset['train'])
        # .astype(np.float(32))
        for i in range(10):
            s = time.time()
            for i in range(nb // ni):
                start = i * ni
                end = (i + 1) * ni
                # int_values = np.arange(start, end, dtype='int64')
                print(f'start insert {start}:{end}')
                data = [vectors[start: end]]
                collection_w.insert(data)
            log.info(collection_w.num_entities)
            log.info(f'{i} insert cost: {time.time() - s}')
```
3. At one run, create collection hangs with timeout 20
```
utility.list_collections()
['disk_3Y7MChs5', 'disk_r2NHiyMw', 'disk_pem1gMUE', 'disk_WjNQUIyW', 'disk_hZbh5k6Q']
```

Milvus logs:
[milvus_logs.tar.gz](https://github.com/milvus-io/milvus/files/8305399/milvus_logs.tar.gz)

 

### Expected Behavior

Maybe create collection successfully, or get the timeout exception

### Anything else?

Pulsar topics stats: https://zilliverse.feishu.cn/docs/doccn34iE0kVvTinUFLqo7ZfNdc

**Comment by yanliang567** [comment]
@ThreadDao  about "However, during this period, the disk of the node was filled many times, causing many pods to be evicted and restarted.", is it a designed chaos by testcase or something wrong by accident?

**Comment by XuanYang-cn** [comment]
@ThreadDao I need more information to identify what's reason for DataNode to fail for the first time:

1. What's the **total size** in GB of data you intend to insert, please add them all.
2. Which node consumes disks? Pulsar? DataNode?
3. What's the retentionInTime and retentionInSize configs of pulsar

**Comment by XuanYang-cn** [comment]
> @ThreadDao about "However, during this period, the disk of the node was filled many times, causing many pods to be evicted and restarted.", is it a designed chaos by testcase or something wrong by accident?

According to pulsar's log, the restarted DataNode subscribed with a different subscirpition name, especially `by-dev-dataNode-38-431891580976889857"`, which consumes nothing, causing all data stalled in backlog, causing no producers able to create or produce(timeout with 30s by default), which stuck the create collection part.

So there're 2 problems here:

1. Why DataNode restarts for the first time
2. DC and DN's unsubscription procedure need to refine.

**Comment by ThreadDao** [comment]
> @ThreadDao about "However, during this period, the disk of the node was filled many times, causing many pods to be evicted and restarted.", is it a designed chaos by testcase or something wrong by accident?

I did this experiment for the first time and didn't design it, it's just what happened in my experiment

**Comment by ThreadDao** [comment]
> 

The reason the DataNode failed the first time was that I increased the disk fill causing the pod to be evicted and restarted
1. I intend to insert as much data as possible. There 6 collections: 
```
collections and num_entities: 'disk_hZbh5k6Q': 10m
 'disk_3Y7MChs5': 10m
 'disk_r2NHiyMw': 10m
 'disk_pem1gMUE': 6m
 'disk_WjNQUIyW': 9m
```
2.  Which node consumes disk? I don't quite understand what you want to know
3. I didn't set retentionInTime and retentionInSize , it should be the default

**Comment by ThreadDao** [comment]
> 

I think the reason for the DN restart is that out of disk pods are evicted and restarted. As for why the disk is not enough because I fill the disk (this is chaos experment)

**Comment by ThreadDao** [comment]
And I re-insert new data into the existed collection 'disk_3Y7MChs5', insert successfully but flush entities hangs

**Comment by ThreadDao** [comment]
> I think the reason for the DN restart is that out of disk pods are evicted and restarted. As for why the disk is not enough because I fill the disk (this is chaos experment)

I think the reason for the DN restart is that out of disk pods are evicted and restarted. As for why the disk is not enough because I fill the disk (this is chaos experment)

**Comment by XuanYang-cn** [comment]
- The problem:
1. DC and DN crashed at the same time
2. DN's old subscriptions are unable to close by DC

- Temporary solution:
Unsubscribe DN's old subscription by hand.

- Real solution:
#15846

**Comment by ThreadDao** [comment]
I found out yesterday that dml_2 topics added a subscription from a new dataNode id-38. It is worth noting that the dataNode and dataCoord did not restart, but all pulsar bookkeeper pods restarted several times.
```
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_2"
"by-dev-dataNode-47-431910158316863489"
"by-dev-dataNode-47-431906376226963457"
"by-dev-dataNode-38-431906376226963457"
```
After i unscribe the "by-dev-dataNode-38-431906376226963457", flush not hangs.

**Comment by ThreadDao** [comment]
When I upgrade the existed Milvus cluster with the new image `master-20220329-e2e34507`, the datanode panic
```
1092 time="2022-03-29T14:40:47Z" level=info msg="[Reconnecting to broker in  102.765569ms]" consumerID=6 name=bwqwv subscription=by-dev-dataNode-95-431906376226963457 topic="pers     istent://public/default/by-dev-rootcoord-dml_2"
1093 panic: runtime error: invalid memory address or nil pointer dereference
1094 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x18b67d2]
1095
1096 goroutine 192 [running]:
1097 github.com/milvus-io/milvus/internal/storage.(*PrimaryKeyStats).UnmarshalJSON(0xc000654ff0, 0xc003638000, 0x2ce61, 0x30000, 0x7f285c3f4488, 0xc000654ff0)
1098     /go/src/github.com/milvus-io/milvus/internal/storage/stats.go:60 +0x172
1099 encoding/json.(*decodeState).object(0xc0005f56b0, 0x2ffc780, 0xc00011a250, 0x16, 0xc0005f56d8, 0xc000971b7b)
1100     /usr/local/go/src/encoding/json/decode.go:609 +0x20a3
1101 encoding/json.(*decodeState).value(0xc0005f56b0, 0x2ffc780, 0xc00011a250, 0x16, 0xc000971b98, 0xb1bdb9)
1102     /usr/local/go/src/encoding/json/decode.go:370 +0x6d
1103 encoding/json.(*decodeState).unmarshal(0xc0005f56b0, 0x2ffc780, 0xc00011a250, 0xc0005f56d8, 0x0)
1104     /usr/local/go/src/encoding/json/decode.go:180 +0x1f4
1105 encoding/json.Unmarshal(0xc003638000, 0x2ce61, 0x30000, 0x2ffc780, 0xc00011a250, 0x31101a0, 0x1)
1106     /usr/local/go/src/encoding/json/decode.go:107 +0x11a
1107 github.com/milvus-io/milvus/internal/storage.(*StatsReader).GetPrimaryKeyStats(0xc000971cd8, 0x0, 0x3, 0xc00090b248)
1108     /go/src/github.com/milvus-io/milvus/internal/storage/stats.go:212 +0x99
1109 github.com/milvus-io/milvus/internal/storage.DeserializeStats(0xc0000359a0, 0x3, 0x4, 0x2, 0x3, 0xc0000359a0, 0x2, 0x4)
1110     /go/src/github.com/milvus-io/milvus/internal/storage/stats.go:229 +0xe6
1111 github.com/milvus-io/milvus/internal/datanode.(*SegmentReplica).initPKBloomFilter(0xc0001a4720, 0xc000657000, 0xc00004f900, 0x3, 0x4, 0x5ff5808bdc80001, 0x0, 0x33e903c)
1112     /go/src/github.com/milvus-io/milvus/internal/datanode/segment_replica.go:470 +0x4cd
1113 github.com/milvus-io/milvus/internal/datanode.(*SegmentReplica).addNormalSegment(0xc0001a4720, 0x5ff09e6647f0d41, 0x5fe745f844c0001, 0x5fe745f844c0002, 0xc000afc0c0, 0x2b, 0     x249f0, 0xc00004f900, 0x3, 0x4, ...)
1114     /go/src/github.com/milvus-io/milvus/internal/datanode/segment_replica.go:373 +0x6e5
1115 github.com/milvus-io/milvus/internal/datanode.(*dataSyncService).initNodes(0xc0009080c0, 0xc0005c7710, 0x368de90, 0xc000436340)
1116     /go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:173 +0x9a5
1117 github.com/milvus-io/milvus/internal/datanode.newDataSyncService(0x368de90, 0xc000639f80, 0xc000a39000, 0x36b2758, 0xc0001a4720, 0x3681210, 0xc000116b80, 0x368e718, 0xc00050     1580, 0xc0005c7710, ...)
1118     /go/src/github.com/milvus-io/milvus/internal/datanode/data_sync_service.go:90 +0x228
1119 github.com/milvus-io/milvus/internal/datanode.(*flowgraphManager).addAndStart(0xc000520660, 0xc000656300, 0xc0005c7710, 0x2, 0x2)
1120     /go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_manager.go:58 +0x7dd
1121 github.com/milvus-io/milvus/internal/datanode.(*DataNode).handlePutEvent(0xc000656300, 0xc000679c80, 0x1, 0x0, 0x0)
1122     /go/src/github.com/milvus-io/milvus/internal/datanode/data_node.go:370 +0x2eb
1123 github.com/milvus-io/milvus/internal/datanode.(*channelEventManager).retryHandlePutEvent(0xc000436300, 0x1, 0xc000afc030, 0x2b, 0x1, 0xc000679c80)
1124     /go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:88 +0x118
1125 github.com/milvus-io/milvus/internal/datanode.(*channelEventManager).Run.func1(0xc000436300)
1126     /go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:70 +0x165
1127 created by github.com/milvus-io/milvus/internal/datanode.(*channelEventManager).Run
1128     /go/src/github.com/milvus-io/milvus/internal/datanode/event_manager.go:64 +0x3f
```
[datanode.log](https://github.com/milvus-io/milvus/files/8372748/datanode.log)

**Comment by xiaofan-luan** [comment]
I

> /assign @xige-16 #15993 seems to break the compatibility

I will fix the issue, pleas open a saperate issue for me~

**Comment by XuanYang-cn** [comment]
> /assign @xige-16 #15993 seems to break the compatibility

Yes, line 211 only supports Int64 as PK
```
 209 // GetInt64Stats returns buffer as PrimaryKeyStats
  210 func (sr *StatsReader) Get PrimaryKeyStats contains statistics data for pk column
A>211     stats := &PrimaryKeyStats{}
  212     err := json.Unmarshal(sr.buffer, &stats)
  213     if err != nil {
  214         return nil, err
  215     }
  216
  217     return stats, nil
  218 }
```

**Comment by ThreadDao** [comment]
> I
> 
> > /assign @xige-16 #15993 seems to break the compatibility
> 
> I will fix the issue, pleas open a saperate issue for me~

Done. #16283

**Comment by ThreadDao** [comment]
I deploy a fresh cluster with image 'master-20220401-5494f3c', after datanode and datacoord restart 3 or 4 times because of chaos disk attack, some subscriptions of the offline datanodes were not successfully unsubscribed:
the dataNode 12 is offline
```
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_2"
"by-dev-dataNode-12-432241213403037697"
"by-dev-dataNode-22-432241213403037697"
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_8"
"by-dev-dataNode-22-432241684777271297"
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_1"
"by-dev-dataNode-12-432241213403037697"
"by-dev-dataNode-22-432245201756425537"
"by-dev-dataNode-22-432241213403037697"
# ./pulsar-admin topics list "public/default"
"persistent://public/default/by-dev-rootcoord-dml_5"
"persistent://public/default/by-dev-rootcoord-dml_6"
"persistent://public/default/by-dev-rootcoord-dml_3"
"persistent://public/default/by-dev-rootcoord-dml_4"
"persistent://public/default/by-dev-rootcoord-dml_7"
"persistent://public/default/by-dev-rootcoord-dml_8"
"persistent://public/default/by-dev-rootcoord-timetick"
"persistent://public/default/by-dev-datacoord-timetick-channel"
"persistent://public/default/by-dev-rootcoord-delta_8"
"persistent://public/default/milvus-operator-topic"
"persistent://public/default/by-dev-rootcoord-delta_6"
"persistent://public/default/by-dev-rootcoord-delta_7"
"persistent://public/default/by-dev-rootcoord-delta_0"
"persistent://public/default/by-dev-rootcoord-delta_1"
"persistent://public/default/by-dev-rootcoord-delta_4"
"persistent://public/default/by-dev-rootcoord-dml_1"
"persistent://public/default/by-dev-rootcoord-delta_5"
"persistent://public/default/by-dev-rootcoord-dml_2"
"persistent://public/default/by-dev-rootcoord-delta_2"
"persistent://public/default/by-dev-rootcoord-delta_3"
"persistent://public/default/by-dev-rootcoord-dml_0"
```

**Comment by congqixia** [comment]
Might be solved by #16946

**Comment by ThreadDao** [comment]
/unassign /assign @XuanYang-cn @congqixia 
Deploy a fresh cluster with image 'master-20220520-d525e955', after datanode and datacoord restart 3 times because of chaos disk attack `chaosd attack disk fill -c 87 -p /tmp`, some subscriptions of the offline datanodes were not successfully unsubscribed:
Obviously datanode 21 has gone offline
```
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_0"
"by-dev-dataNode-42-433407184247455745"
"by-dev-dataNode-42-433406364697231361"
"by-dev-dataNode-21-433407184247455745"
"by-dev-dataNode-21-433406364697231361"
#
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_1"
"by-dev-dataNode-42-433407184247455745"
"by-dev-dataNode-42-433406364697231361"
#
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_2"
"by-dev-dataNode-42-433406364697231361"
#
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_3"
"by-dev-dataNode-42-433406364697231361"
#
# ./pulsar-admin topics subscriptions "persistent://public/default/by-dev-rootcoord-dml_4"
#
```

**Comment by XuanYang-cn** [comment]
From my tests, it's not because disk-stress. The recovery of disk stress is simple and straightforward, both down and both up.

I guess the frequent restart is the cause, try to reproduce now

**Comment by XuanYang-cn** [comment]
```
[2022/05/25 17:30:04.711 +08:00] [WARN] [channel_manager.go:577] ["fail to set channels to release for watch failure ACKs"] [nodeID=12] ["channel name"=by-dev-rootcoord-dml_1_433447435450912065v1] [error="fail to find matching nodeID: 12 with channelName: by-dev-rootcoord-dml_1_433447435450912065v1"]
```

When unable to set to release, DataCoord should directly clear the subscription

**Comment by xiaofan-luan** [comment]
> ```
>  85 [2022/05/25 17:28:08.734 +08:00] [INFO] [channel_manager.go:327] ["register node"] ["registered node"=9] 
> [updates="[\"{type=Delete,nodeID=7,channels=\\\"[by-dev-rootcoord-dml    _0_433447435450912065v0]\\\"}\",\"
> {type=Add,nodeID=9,channels=\\\"[by-dev-rootcoord-dml_0_433447435450912065v0]\\\"}\"]"]
> ```
> 
> Register is using `AvgAssignPolicy`, which will delete some channels from an old DataNode; and we don't guarantee the delete. Once the register success, and the old DataNode crashed, the subscribed channel won't be deleted forever.

Does that means we simply missed deregister task if datanode and datacoord down at the same time? 
I guess the easiest way to fix is scan all the channel and trigger deregister if some of the channel is on dead datanode

**Comment by XuanYang-cn** [comment]
> 
> 
> Does that means we simply missed deregister task if datanode and datacoord down at the same time?
> I guess the easiest way to fix is scan all the channel and trigger deregister if some of the channel is on dead datanode

@xiaofan-luan This's not the problem. In my test, DataNode restarts several times but DataCoord stays healthy.

**Because the detection of Node offline may take a while, and the detection of a new Node online happens immediately**, when a new Node is online, DataCoord thinks there're 2 alive Nodes available, and starts to balance the channels. This balance is triggered by `register node`. And we're using `AvgAssignPolicy`, which means balance the channels equally among all the online Nodes. 
- This balance deletes some channels from the old Node, adds channels into the registered new Node

It works perfectly fine with normal cases. The problem is , this is a Delete event, the old DataNode needs to release the balanced flow graph.

If the old DataNode crashed during the release, this channel subscription will never be released because:
- DataCoord doesn't guarantee delete
- When DataCoord finally detects the old DataNode is offline, This channel has already been moved to the New registered Node, it's not recorded on the old crashed Node, so nothing needs to be cleaned from DC's point of review.

**Comment by XuanYang-cn** [comment]
Here's a workable solution @xiaofan-luan . We can make un-subscription more powerful in DataCoord:

1. To deal with offline nodes problem,  whenever DC detects a DataNode offline, it'll use `pulsarctl` to find all subscriptions by the offline DataNode, and unsubscribe all the offline Nodes' subscriptions.
2. To deal with online Nodes  load balance `delete events release-failure`, DataCoord can check delete events after 5 minutes via `pulsarctl` to verify if the the subscription is closed as expected. If not, just help to unsubscribe it.

**Comment by xiaofan-luan** [comment]
> * DataCoord doesn't guarantee delete

as you just mentioned, "DataCoord doesn't guarantee delete"
If just want a quick fix about this, should we just wait until channel unregister is handled before finish the assign?

**Comment by XuanYang-cn** [comment]
> > * DataCoord doesn't guarantee delete
> 
> as you just mentioned, "DataCoord doesn't guarantee delete" If just want a quick fix about this, should we just wait until channel unregister is handled before finish the assign?

Yes we could make DataCoord guarantee the deletion of a channel.

DC put ToRelease -> DN release and delete this key

**Comment by XuanYang-cn** [comment]
![origin_img_v2_9a4351ee-f3b3-4c3d-9795-2aad47fb8a8g](https://user-images.githubusercontent.com/51370125/170620243-043e01d3-5f00-43ea-8c8e-3c8ff63945d4.jpg)

**Comment by xiaofan-luan** [comment]
> ![origin_img_v2_9a4351ee-f3b3-4c3d-9795-2aad47fb8a8g](https://user-images.githubusercontent.com/51370125/170620243-043e01d3-5f00-43ea-8c8e-3c8ff63945d4.jpg)

Overall looks good to me, let's ship it!

**Comment by XuanYang-cn** [comment]
Tested locally with #17293, no reproducing

![image](https://user-images.githubusercontent.com/51370125/171121823-e7664247-5657-4da3-acb7-d85bd922a170.png)

**Comment by ThreadDao** [comment]
Verified fixed on image `master-20220606-9f786dd7`
